internal/querynode/shard_node_detector.go

```diff
@@ -132,31 +135,93 @@ func (nd *etcdShardNodeDetector) watch(ch clientv3.WatchChan, collectionID, repl
 			return
 		case evt, ok := <-ch:
 			if !ok {
-				log.Warn("event ch closed")
+				log.Warn("watch channel closed, retry...")
+				var watchCh clientv3.WatchChan
+				var ok bool
+				watchCh, ok, revision = nd.rewatch(collectionID, replicaID, revision)
+				if !ok {
+					// detector closed
+					return
+				}
+				nd.wg.Add(1)
+				go nd.watch(watchCh, collectionID, replicaID, revision)
 				return
 			}
+			revision = evt.Header.GetRevision()
 			if err := evt.Err(); err != nil {
 				if err == v3rpc.ErrCompacted {
-					ctx, cancel := context.WithCancel(context.Background())
-					watchCh := nd.client.Watch(ctx, nd.path, clientv3.WithPrefix())
-					go nd.cancelClose(cancel)
+					watchCh, ok, revision := nd.rewatch(collectionID, replicaID, common.LatestRevision)
+					if !ok {
+						// detector closed
+						return
+					}
 					nd.wg.Add(1)
-					go nd.watch(watchCh, collectionID, replicaID)
+					go nd.watch(watchCh, collectionID, replicaID, revision)
 					return
 				}
 				log.Error("failed to handle watch node error", zap.Error(err))
 				panic(err)
 			}
-			for _, e := range evt.Events {
-				switch e.Type {
-				case mvccpb.PUT:
-					nd.handlePutEvent(e, collectionID, replicaID)
-				case mvccpb.DELETE:
-					nd.handleDelEvent(e, collectionID, replicaID)
+			nd.handleEvt(evt, collectionID, replicaID)
+		}
+	}
+}
+
+func (nd *etcdShardNodeDetector) handleEvt(evt clientv3.WatchResponse, collectionID, replicaID int64) {
+	for _, e := range evt.Events {
+		switch e.Type {
+		case mvccpb.PUT:
+			nd.handlePutEvent(e, collectionID, replicaID)
+		case mvccpb.DELETE:
+			nd.handleDelEvent(e, collectionID, replicaID)
+		}
+	}
+}
+
+func (nd *etcdShardNodeDetector) rewatch(collectionID, replicaID, rev int64) (ch clientv3.WatchChan, ok bool, revision int64) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	revision = rev
+	err := retry.Do(ctx, func() error {
+		if revision != common.LatestRevision {
+			ch = nd.client.Watch(ctx, nd.path, clientv3.WithPrefix(), clientv3.WithRev(revision+1))
+		} else {
+			ch = nd.client.Watch(ctx, nd.path, clientv3.WithPrefix())
+		}
+		select {
+		case <-nd.closeCh:
+			return retry.Unrecoverable(errors.New("detector closed"))
+
+		case evt, ok := <-ch:
+			if !ok {
```

Comment:
We'd better check the minimum CompactRevision value from WatchChan and ensure the watched revision is not compacted.

Reply:
if watched revision is compacted, the client will receive ErrCompacted

Reply:
If watched revision is compacted, a response with `ErrCompacted` will be received. Any internally, ETCD client will update the latest revision for each event received.